agent/lib_aws_sdk_php.c

```diff
@@ -295,6 +302,194 @@ void nr_lib_aws_sdk_php_sqs_parse_queueurl(
   cloud_attrs->cloud_region = region;
 }
 
+void nr_lib_aws_sdk_php_lambda_handle(nr_segment_t* auto_segment,
+                                      char* command_name_string,
+                                      size_t command_name_len,
+                                      NR_EXECUTE_PROTO) {
+  nr_segment_t* external_segment = NULL;
+  zval** retval_ptr = NR_GET_RETURN_VALUE_PTR;
+
+  nr_segment_cloud_attrs_t cloud_attrs = {
+      .cloud_platform = "aws_lambda"
+  };
+
+  if (NULL == auto_segment) {
+    return;
+  }
+
+  if (NULL == command_name_string || 0 == command_name_len) {
+    return;
+  }
+
+  if (NULL == *retval_ptr) {
+    /* Do not instrument when an exception has happened */
+    return;
+  }
+
+#define AWS_COMMAND_IS(CMD) \
+  (command_name_len == (sizeof(CMD) - 1) && nr_streq(CMD, command_name_string))
+
+  /* Determine if we instrument this command. */
+  if (AWS_COMMAND_IS("invoke")) {
+    /* reconstruct the ARN */
+    nr_aws_sdk_lambda_client_invoke_parse_args(NR_EXECUTE_ORIG_ARGS, &cloud_attrs);
+  } else {
+    return;
+  }
+#undef AWS_COMMAND_IS
+
+  /*
+   * By this point, it's been determined that this call will be instrumented so
+   * only create the segment now, grab the parent segment start time, add our
+   * special segment attributes/metrics then close the newly created segment.
+   */
+  external_segment = nr_segment_start(NRPRG(txn), NULL, NULL);
+  if (NULL == external_segment) {
+    nr_free(cloud_attrs.cloud_resource_id);
+    return;
+  }
+  /* re-use start time from auto_segment started in func_begin */
+  external_segment->start_time = auto_segment->start_time;
+  cloud_attrs.aws_operation = command_name_string;
+
+  /* end the segment */
+  nr_segment_traces_add_cloud_attributes(external_segment, &cloud_attrs);
+  nr_segment_external_params_t external_params = {.library = "aws_sdk"};
+  zval* data = nr_php_get_zval_object_property(*retval_ptr, "data");
+  if (nr_php_is_zval_valid_array(data)) {
+    zval* status_code = nr_php_zend_hash_find(Z_ARRVAL_P(data), "StatusCode");
+    if (nr_php_is_zval_valid_integer(status_code)) {
+      external_params.status = Z_LVAL_P(status_code);
+    }
+    zval* metadata = nr_php_zend_hash_find(Z_ARRVAL_P(data), "@metadata");
+    if (NULL != metadata  && IS_REFERENCE == Z_TYPE_P(metadata)) {
+      metadata = Z_REFVAL_P(metadata);
+    }
+    if (nr_php_is_zval_valid_array(metadata)) {
+      zval* uri = nr_php_zend_hash_find(Z_ARRVAL_P(metadata), "effectiveUri");
+      if (nr_php_is_zval_valid_string(uri)) {
```

Comment:
Is an empty string acceptable, or should this be checking for a valid, nonempty string instead with `nr_php_is_zval_non_empty_string`?

Reply:
If it is empty, there is no harm in attaching it to the external params, but also no gain. added a check in [b72ebc1](https://github.com/newrelic/newrelic-php-agent/pull/1023/commits/b72ebc1104e0bf871dca57a2a458e488e1962505)

Reply:
instead of `if (nr_php_is_zval_valid_string(uri) && !nr_strempty(Z_STRVAL_P(uri)))`

Just use `nr_php_is_zval_non_empty_string` which does both of those things.

Reply:
changed in [50aaa3c](https://github.com/newrelic/newrelic-php-agent/pull/1023/commits/50aaa3c55da96aaa0b34449d87ec3623b33c908a)